src/ir/auth_logic/lowering_ast_datalog.h

```diff
@@ -0,0 +1,319 @@
+    std::vector<DLIRAssertion> pred_list = {
+        SpokenCanActAsToDLIR(speaker, canActAs)};
+    return std::make_pair(CanActAsToDLIR(canActAs), pred_list);
+  }
+
+  // The second return value represents 0 or 1 newly generated rules, so an
+  // option might seem more intuitive. However, the interface that consumes
+  // this needs to construct a vector anyway, so a vector is used in the
+  // return type.
+  std::pair<Predicate, std::vector<DLIRAssertion>> BaseFactToDLIR(
+      const Principal& speaker, const BaseFact& base_fact) {
+    return std::visit(
+        [this, &speaker](auto value) {
+          return BaseFactToDLIRInner(speaker, value);
+        },
+        base_fact.GetValue());
+  }
+
+  // This can result in 0 or more new rules because the translation of
+  // nested canSayFacts might result in more than 1 rule.
+  std::pair<Predicate, std::vector<DLIRAssertion>> FactToDLIR(
+      const Principal& speaker, const Fact& fact);
+
+  std::vector<DLIRAssertion> SingleSaysAssertionToDLIR(
+      const Principal& speaker, const Assertion& assertion);
+
+  std::vector<DLIRAssertion> SaysAssertionToDLIR(
+      const SaysAssertion& says_assertion) {
+    std::vector<DLIRAssertion> ret = {};
+    for (const Assertion& assertion : says_assertion.assertions()) {
+      std::vector<DLIRAssertion> single_translation =
+          SingleSaysAssertionToDLIR(says_assertion.principal(), assertion);
+      ret.insert(ret.end(), std::make_move_iterator(single_translation.begin()),
+                 std::make_move_iterator(single_translation.end()));
+    }
+    return ret;
+  }
+
+  std::vector<DLIRAssertion> SaysAssertionsToDLIR(
+      const std::vector<SaysAssertion>& says_assertions) {
+    std::vector<DLIRAssertion> ret = {};
+    for (const SaysAssertion& says_assertion : says_assertions) {
+      auto single_translation = SaysAssertionToDLIR(says_assertion);
+      ret.insert(ret.end(), std::make_move_iterator(single_translation.begin()),
+                 std::make_move_iterator(single_translation.end()));
+    }
+    return ret;
+  }
+
+  // Queries are like predicates with arity 0. Souffle does not have predicates
+  // with arity 0, so we represent them as having one argument which is a
+  // constant.
+  static inline Predicate kDummyPredicate =
+      Predicate("grounded_dummy", {"dummy_var"}, kPositive);
+
+  DLIRAssertion QueryToDLIR(const Query& query) {
+    // The assertions that are normally generated during the translation
+    // from facts to dlir facts are not used for queries.
+    auto [main_pred, not_used] = FactToDLIR(query.principal(), query.fact());
+    main_pred = PushPrincipal("says_", query.principal(), main_pred);
+    Predicate lhs(query.name(), {"dummy_var"}, kPositive);
+    return DLIRAssertion(DLIRCondAssertion(lhs, {main_pred, kDummyPredicate}));
+  }
+
+  std::vector<DLIRAssertion> QueriesToDLIR(const std::vector<Query>& queries) {
+    std::vector<DLIRAssertion> ret = {};
```

Comment:
`ret.reserve(queries.size());` will avoid reallocations within the loop.

Reply:
thanks. I learned something new. fixed.